tests/Unit/DataStructures/VectorImplTestHelper.hpp

```diff
@@ -161,5 +161,358 @@ auto remove_nth(const std::tuple<ValueTypes...>& tup) noexcept {
+    CHECK(not move_constructed.is_owning());
+  }
+  SECTION("Check movement acts appropriately on both source and target refs") {
+    VectorType ref_original_vector;
+    ref_original_vector.set_data_ref(make_not_null(&original_vector));
+    VectorType generated_vector(size);
+    fill_with_random_values(make_not_null(&generated_vector),
+                            make_not_null(&gen), make_not_null(&dist));
+    VectorType generated_vector_copy = generated_vector;
+    ref_original_vector = std::move(generated_vector);
+    // clang-tidy : Intentionally testing use after move
+    CHECK(original_vector != generated_vector);  // NOLINT
+    CHECK(original_vector == generated_vector_copy);
+// Intentionally testing self-move
+#ifdef __clang__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wself-move"
+#endif  // defined(__clang__)
+    ref_original_vector = std::move(ref_original_vector);
+#ifdef __clang__
+#pragma GCC diagnostic pop
+#endif  // defined(__clang__)
+    CHECK(original_vector == generated_vector_copy);
+    VectorType data_check_vector;
+    // clang-tidy: false positive, used after it was moved
+    data_check_vector = ref_original_vector;  // NOLINT
+    CHECK(data_check_vector == generated_vector_copy);
+  }
+  SECTION("Check math affects both data vectors which share a ref") {
+    auto generated_val1 = make_with_random_values<ValType>(
+             make_not_null(&gen), make_not_null(&dist)),
+         generated_val2 = make_with_random_values<ValType>(
+             make_not_null(&gen), make_not_null(&dist));
+    auto sum_generated_vals = generated_val1 + generated_val2;
+    VectorType sharing_vector{size, generated_val1};
+    VectorType owning_vector{size, generated_val2};
+    sharing_vector.set_data_ref(make_not_null(&owning_vector));
+    sharing_vector = sharing_vector + generated_val1;
+    detail::check_vectors(owning_vector, sum_generated_vals);
+    detail::check_vectors(sharing_vector, sum_generated_vals);
+  }
+}
+
+/// \ingroup TestingFrameworkGroup
+/// \brief test that `VectorType` appropriately errors out when assigned to
+/// another `VectorType` with the wrong size that is non-owning.
+///
+/// \details a calling function should be an `ASSERTION_TEST()` and check for
+/// the string "Must copy into same size".
+template <typename VectorType,
+          typename ValType = typename VectorType::ElementType>
+void vector_ref_test_size_error(
+    typename tt::get_fundamental_type_t<ValType> low =
+        typename tt::get_fundamental_type_t<ValType>{-100},
+    typename tt::get_fundamental_type_t<ValType> high =
+        typename tt::get_fundamental_type_t<ValType>{100}) noexcept {
+  MAKE_GENERATOR(gen);
+  UniformCustomDistribution<typename tt::get_fundamental_type_t<ValType>> dist{
+      low, high};
+  UniformCustomDistribution<size_t> sdist{2, 20};
+
+  size_t size = sdist(gen);
+  VectorType generated_vector{size, 0};
```

Comment:
` ... make_with_random_values`?

(note to self: can't be const because reference things)